File: app/src/main/java/com/example/androidproject/views/EditUserProfileFragment.java
package com.example.androidproject.views;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;

import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.fragment.app.Fragment;
import androidx.navigation.Navigation;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.androidproject.R;
import com.example.androidproject.model.CloudinaryModel;
import com.example.androidproject.model.User;
import com.example.androidproject.databinding.FragmentEditUserProfileBinding;
import com.example.androidproject.repositories.UserRepository;
import com.squareup.picasso.Picasso;

public class EditUserProfileFragment extends Fragment {

    FragmentEditUserProfileBinding binding;
    ActivityResultLauncher<Void> cameraLauncher;
    Boolean isImgSelected = false;
    User user;

    public EditUserProfileFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        binding = FragmentEditUserProfileBinding.inflate(inflater, container, false);
        View view = binding.getRoot();

        user = getArguments().getParcelable("User");

        if (user != null) {
            binding.usernameEditTp.setText(user.getUsername());
            binding.mailTp.setText(user.getMail());
        }

        cameraLauncher = registerForActivityResult(new ActivityResultContracts.TakePicturePreview(), new ActivityResultCallback<Bitmap>() {
            @Override
            public void onActivityResult(Bitmap result) {
                if (result != null) {
                    binding.userImg.setImageBitmap(result);
                    isImgSelected = true;
                }
            }
        });

        if (user != null && user.getImgUrl() != null && !user.getImgUrl().isEmpty()) {
            Picasso.get().load(user.getImgUrl()).placeholder(R.drawable.profile).into(binding.userImg);
        } else {
            binding.userImg.setImageResource(R.drawable.profile);
        }

        onPhotoClick();
        onSave(view);

        return view;
    }

    private void onPhotoClick() {
        binding.userImg.setOnClickListener(v -> cameraLauncher.launch(null));
    }

    private void uploadImg(User user, Context context, View view) {
        binding.userImg.setDrawingCacheEnabled(true);
        binding.userImg.buildDrawingCache();
        Bitmap bitmap = ((BitmapDrawable) binding.userImg.getDrawable()).getBitmap();

        new CloudinaryModel().uploadImage(bitmap, user.getMail(), url -> {
            if (url != null && !url.isEmpty()) {
                user.setImgUrl(url);
            }
            saveUserNewData(user, view);
            return null;
        }, error -> {
            // Optional: show a Toast or log the error
            saveUserNewData(user, view);
            return null;
        });
    }

    public void onSave(View view) {
        binding.saveEditBtn.setOnClickListener(v -> {
            User editedUser = new User(user.getPassword(), binding.mailTp.getText().toString(), binding.usernameEditTp.getText().toString(), user.getUid(), user.getImgUrl());
            if (isImgSelected) {
                uploadImg(editedUser, requireContext(), view);
            } else {
                saveUserNewData(editedUser, view);
            }
        });
    }

    private void saveUserNewData(User editedUser, View view) {
        UserRepository.instance.updateUser(editedUser, unused -> Navigation.findNavController(view).navigate(R.id.action_editUserProfileFragment_to_profileFragment));
    }
}

File: app/src/main/java/com/example/androidproject/views/EditUserProfileFragment.kt
package com.example.androidproject.views

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentTransaction
import com.example.androidproject.R
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import java.util.HashMap

class EditUserProfileFragment : Fragment() {

    private lateinit var emailEditText: EditText
    private lateinit var usernameEditText: EditText
    private lateinit var saveButton: Button

    private lateinit var mAuth: FirebaseAuth
    private lateinit var db: FirebaseFirestore

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_edit_user_profile, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // Initialize Firebase
        mAuth = FirebaseAuth.getInstance()
        db = FirebaseFirestore.getInstance()

        // Initialize UI elements
        emailEditText = view.findViewById(R.id.mail_tp)
        usernameEditText = view.findViewById(R.id.username_edit_tp)
        saveButton = view.findViewById(R.id.save_edit_btn)

        // Load user data
        loadUserData()

        // Set up click listener
        saveButton.setOnClickListener { saveUserData() }
    }

    private fun loadUserData() {
        val userId = mAuth.currentUser?.uid ?: return

        // Set email from Firebase Auth
        emailEditText.setText(mAuth.currentUser?.email)

        // Get username from Firestore
        db.collection("users").document(userId)
            .get()
            .addOnSuccessListener { documentSnapshot ->
                if (documentSnapshot.exists()) {
                    val username = documentSnapshot.getString("username")
                    if (username != null) {
                        usernameEditText.setText(username)
                    }
                }
            }
            .addOnFailureListener { e ->
                Toast.makeText(context, "Error loading user data: ${e.message}", Toast.LENGTH_SHORT).show()
            }
    }

    private fun saveUserData() {
        val username = usernameEditText.text.toString().trim()

        if (username.isEmpty()) {
            usernameEditText.error = "Username is required"
            return
        }

        val userId = mAuth.currentUser?.uid ?: return

        val userData = HashMap<String, Any>()
        userData["username"] = username

        db.collection("users").document(userId)
            .update(userData)
            .addOnSuccessListener {
                Toast.makeText(context, "Profile updated successfully", Toast.LENGTH_SHORT).show()

                // Navigate back to profile fragment
                val transaction = parentFragmentManager.beginTransaction()
                transaction.replace(R.id.fragmentContainer, ProfileFragment())
                transaction.commit()
            }
            .addOnFailureListener { e ->
                Toast.makeText(context, "Error updating profile: ${e.message}", Toast.LENGTH_SHORT).show()
            }
    }
}
